Dataset revision endpoint implementation

A-Artemis opened the pull request:
<!-- Contributing guide: https://github.com/open-edge-platform/training_extensions/blob/develop/CONTRIBUTING.md -->

### Summary

Adds the implementation for the Dataset Revision endpoints.
- get
- get binary/thumbnail
- list
- delete

BENCHMARK 1: Storage Format Comparison

Storage Format       Size            Export Time     Savings
#----------------------------------------------------------------------
Uncompressed         2.93 MB         0.210s                
Zip Compressed       2.93 MB         0.374s          0.2%

BENCHMARK 2: Decompression Overhead

Extraction time: 0.175s
Load from uncompressed: 0.165s
Total overhead (extract + load): 0.340s

BENCHMARK 3: Query Performance

Method 1: Direct Polars Query
  Select all                               2.84ms (100 rows)
  Filter by subset                         3.29ms (80 rows)
  Paginate (limit 10, offset 0)            1.77ms (10 rows)
  Paginate (limit 10, offset 50)           1.62ms (10 rows)

Method 2: Datumaro Dataset Query
  Dataset load time                        142.69ms
  Dataset length                           100 items

#--------------------------------------------------------------------------------
Query Performance Summary:
  Average Polars query time:               2.38ms
  Datumaro load time:                      142.69ms

Polars is significantly faster for querying (59.9x faster than Datumaro load)

### How to test

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [x] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

A-Artemis committed:
Dataset revision endpoint implementation

A-Artemis committed:
Reimplemented tests

A-Artemis committed:
test fix

leoll2 commented on application/backend/app/services/dataset_service.py:
```diff
@@ -456,6 +459,184 @@ def delete_dataset_revision_files(self, project_id: UUID, revision_id: UUID) ->
+        # Otherwise it's uncompressed (though this shouldn't happen with current implementation)
+        return revision_path / "data.parquet"
+
+    def list_dataset_revision_items(
```
I suggest to move the new logic to a separate service `DatasetRevisionService`, it is easier to maintain and there is not really much in common with the other methods of `DatasetService`.

leoll2 commented on application/backend/app/api/routers/dataset_revisions.py:
```diff
@@ -33,15 +36,45 @@
+            format=DatasetItemFormat(Path(item_data["image"]).suffix.lstrip(".")),
+            width=item_data["image_info"]["width"],
+            height=item_data["image_info"]["height"],
+            size=0,  # Not available in parquet
```
It's generally better to use `None` to represent "unknown" instead of 0. In this case, I'd suggest to define a new view model `DatasetRevisionItemView`, similar to `DatasetItemView` but without `size` and `source_id`.

leoll2 commented on application/backend/app/api/routers/dataset_revisions.py:
```diff
@@ -89,13 +142,18 @@ def get_dataset_revision_item_binary(
 ) -> FileResponse:
     """Get the thumbnail of an item in the dataset revision"""
-    raise NotImplementedError
+    binary_path = dataset_service.get_dataset_revision_item_binary_path(
```
Returning the full image as thumbnail is not useful; we can address the thumbnail problem in another PR, for example we can generate the thumbnail lazily and cache it (next to the dataset revision) for subsequent requests.

leoll2 commented on application/backend/app/services/dataset_service.py:
```diff
@@ -456,6 +459,184 @@ def delete_dataset_revision_files(self, project_id: UUID, revision_id: UUID) ->
+
+        # When as_zip=True, datumaro creates dataset.zip inside the output directory
+        zip_path = revision_path / "dataset.zip"
+        if zip_path.exists():
```
Before checking if the zip path exists, you could check if `data.parquet` already exists.

Pseudocode:
```
if parquet file exists:
  return parquet file path
elif zip file exist:
  uncompress
  return parquet file path
else:
  raise exception
```

leoll2 commented on application/backend/app/services/dataset_service.py:
```diff
@@ -456,6 +459,184 @@ def delete_dataset_revision_files(self, project_id: UUID, revision_id: UUID) ->
+            return items, total_count
+        finally:
+            # Clean up temp directory if we extracted a zip
+            if "revision_" in str(parquet_path.parent.name) and parquet_path.parent.parent.name != "dataset_revisions":
```
Cleanup should be ideally done in the same context where you allocate the resource (i.e. close to where you create the temp directory)

A-Artemis committed:
Refactor dataset revision endpoint to use new DatasetRevisionService and update schemas

leoll2 commented on application/backend/app/api/schemas/dataset_item.py:
```diff
@@ -87,7 +111,7 @@ class DatasetItemsWithPagination(BaseModel):
     """
 
-    items: list[DatasetItemView]
+    items: list[DatasetItemView | DatasetItemRevisionView]
```
Please define a separate model `DatasetRevisionItemsWithPagination` for the revision endpoints, so that the UI doesn't have to deal with the union type (annoying for type-checking).

leoll2 commented on application/backend/app/services/dataset_revision_service.py:
```diff
@@ -0,0 +1,282 @@
+        revision_db = revision_repo.get_by_id(str(revision_id))
+        if revision_db:
+            revision_db.files_deleted = True
+            revision_repo.save(revision_db)
```
To prevent race conditions, it's better to update the DB first (mark the revision as deleted -> inaccessible) and delete the files immediately after.

leoll2 commented on application/backend/app/services/dataset_revision_service.py:
```diff
@@ -0,0 +1,282 @@
+        Returns:
+            Tuple of (list of items as dicts, total count).
+        """
+        parquet_path, is_temp = self._get_revision_parquet_path(project_id, revision_id)
```
In the current implementation, `list_dataset_revision_items` would uncompress the dataset everytime it is called. This doesn't play well with pagination: if e.g. you call the endpoints 3 times to get 150 items in pages of 50, then you uncompress the same dataset 3 times in a row, which is clearly inefficient.

Assuming that storing the dataset revisions in a compressed zip format would significantly reduce the disk usage (to be verified empirically), then we need a better approach to uncompress the dataset when needed and keep it available, uncompressed, for subsequent requests in a short time frame. For example, we could implement a mechanism that caches the uncompressed datasets somewhere, and automatically deletes them after some time (e.g. a few hours) or when the disk space gets low.

Anyway, to make a proper decision, we first need answers to the questions in the ticket.

A-Artemis committed:
Merge branch 'develop' of github.com:open-edge-platform/training_extensions into aurelien/dataset-revision-endpoint-implementation

A-Artemis committed:
Export DatasetRevisions uncompressed (as_zip=False)
Added new DatasetRevisionItemsWithPagination
Simplified DatasetRevision save/delete

A-Artemis committed:
Fixed tests

A-Artemis committed:
Fixed mypy issues

A-Artemis committed:
test fixes

A-Artemis committed:
test fixes

leoll2 commented on application/backend/app/api/routers/dataset_revisions.py:
```diff
@@ -33,33 +40,74 @@
+            height=item_data["image_info"]["height"],
+            subset=DatasetItemSubset(item_data["subset"]),
+        )
+        items.append(item_view)
```
Ideally, this logic should be in `DatasetRevisionService.list_dataset_revision_items`, which should return an object of type `app.models.dataset_item_revision.DatasetRevisionItem` (the domain entity) instead of a raw dictionary. The endpoint should then convert to the view type with `DatasetItemRevisionView.model_validate(dataset_revision_item, from_attributes=True)`.

For reference, you can check the implementation of `list_dataset_items` in `api/routers/datasets.py`.

leoll2 reviewed (commented):
LGTM, only one suggestion to improve the service interface through the use of business-layer entities, like the other services do.

A-Artemis committed:
Refactor dataset revision models and endpoints
